packages/@webex/plugin-meetings/src/reachability/clusterReachability.ts

```diff
@@ -0,0 +1,325 @@
+    try {
+      const config = this.buildPeerConnectionConfig(clusterInfo);
+
+      console.log(`marcin: config=${JSON.stringify(config)}`);
+      const peerConnection = new RTCPeerConnection(config);
+
+      return peerConnection;
+    } catch (peerConnectionError) {
+      LoggerProxy.logger.log(
+        `Reachability:index#createPeerConnection --> Error creating peerConnection: ${peerConnectionError}`
+      );
+
+      return undefined;
+    }
+  }
+
+  getResult() {
+    return this.result;
+  }
+
+  isUnreachable(protocol: 'udp' | 'tcp') {
+    if (protocol === 'udp') {
+      return this.numUdpUrls > 0 && this.result.udp.reachable !== 'true';
+    }
+
+    return this.numTcpUrls > 0 && this.result.tcp.reachable !== 'true';
+  }
+
+  private closePeerConnection() {
+    if (this.pc) {
+      this.pc.onicecandidate = null;
+      this.pc.onicegatheringstatechange = null;
+      this.pc.close();
+    }
+  }
+
+  private finishReachabilityCheck() {
+    this.defer.resolve();
+  }
+
+  /**
+   * Adds public IP (client media IPs)
+   * @param {string} protocol
+   * @param {string} publicIP
+   * @returns {void}
+   */
+  protected addPublicIP(protocol: 'udp' | 'tcp', publicIP?: string | null) {
+    const {CLOSED} = CONNECTION_STATE;
+
+    if (this.pc?.connectionState === CLOSED) {
+      LoggerProxy.logger.log(
+        `Reachability:index#addPublicIP --> Attempting to set publicIP of ${publicIP} on closed peerConnection.`
+      );
+    }
+
+    const result = this.result[protocol];
+
+    if (publicIP) {
+      if (result.clientMediaIPs) {
+        if (!result.clientMediaIPs.includes(publicIP)) {
+          result.clientMediaIPs.push(publicIP);
+        }
+      } else {
+        result.clientMediaIPs = [publicIP];
+      }
+    }
+  }
+
+  private registerIceGatheringStateChangeListener() {
+    this.pc.onicegatheringstatechange = () => {
+      const {COMPLETE} = ICE_GATHERING_STATE;
+
+      if (this.pc.iceConnectionState === COMPLETE) {
+        console.log(
+          `marcin: ${this.name}: this.pc.iceConnectionState=${this.pc.iceConnectionState}`
+        );
+        this.closePeerConnection();
+        this.finishReachabilityCheck();
+      }
+    };
+  }
+
+  private haveWeGotAllResults(): boolean {
+    const expecting = {
+      udp: this.numUdpUrls > 0,
+      tcp: this.numTcpUrls > 0,
+    };
+
+    const gotResult = {
+      udp: this.result.udp.reachable === 'true',
+      tcp: this.result.tcp.reachable === 'true',
+    };
+
+    return ['udp', 'tcp'].every((protocol) => expecting[protocol] === gotResult[protocol]);
+  }
+
+  private registerIceCandidateListener() {
+    this.pc.onicecandidate = (e) => {
+      const CANDIDATE_TYPES = {
+        SERVER_REFLEXIVE: 'srflx',
+        RELAY: 'relay',
+      };
+
+      if (e.candidate) {
+        if (e.candidate.type !== 'host') {
+          console.log('marcin: e.candidate', e.candidate);
+        }
+        if (e.candidate.type === CANDIDATE_TYPES.SERVER_REFLEXIVE) {
```

Comment:
these 2 ifs are the same, maybe we can abstract it into a function, should be easier to test and reduces duplication

Reply:
also does reflexive = UDP, and relay = TCP?

Reply:
> also does reflexive = UDP, and relay = TCP?

yes, it's because of how we do it - we use Linus as STUN server to check UDP reachability, so they generate reflexive candidates. Webrtc only supports talking to STUN servers over UDP. For TCP we use Linus pretending to be a TURN server (that generates RELAY candidates) - see `buildPeerConnectionConfig()` where we force the TURN connection to be done only over TCP

Reply:
moved the common code from the 2 if blocks into `storeLatencyResult()`